File: web/bodyimport.js
/**
 * Einlesen von Waagendaten aus einer geteilten Datei.
 *
 * Wortgleich zur Android-Fassung (android/…/data/BodyImport.kt) — beide müssen dieselbe
 * Datei gleich verstehen, sonst steht der Bestand auf einem der Geräte anders drin.
 *
 * Hintergrund: Ein direkter Zugriff auf die FitDays-App ist nicht möglich, Android kapselt
 * Apps gegeneinander ab. Klappt der Weg über Health Connect nicht, bleibt der Export —
 * und weil jede Waagen-App ein eigenes Format schreibt, ist dieser Leser bewusst großzügig.
 */

/** Spaltennamen, die je Wert in Frage kommen — in mehreren Sprachen. */
const WEIGHT = ['gewicht', 'weight', 'masse', 'kg'];
const DATE = ['datum', 'date', 'zeit', 'time', 'gemessen', 'measured'];
const FAT = ['körperfett', 'koerperfett', 'fett', 'body fat', 'bodyfat', 'fat'];
const LEAN = ['magermasse', 'muskel', 'muscle', 'lean'];
const WATER = ['wasser', 'water'];
const BONE = ['knochen', 'bone'];
const BASAL = ['grundumsatz', 'bmr', 'basal', 'kcal'];

const DAY_MS = 86400000;

/** Erkennt das Trennzeichen an der Kopfzeile. */
function detectSeparator(header) {
  return [';', '\t', ','].reduce((best, sep) => {
    const count = header.split(sep).length - 1;
    const bestCount = header.split(best).length - 1;
    return count > bestCount ? sep : best;
  }, ';');
}

const splitLine = (line, sep) => line.split(sep).map((c) => c.trim().replace(/^"|"$/g, ''));

const indexOfAny = (header, needles) =>
  header.findIndex((cell) => needles.some((n) => cell.includes(n)));

/** Zahl aus einer Zelle — akzeptiert Punkt wie Komma und ignoriert Einheiten. */
function toNumber(cell) {
  if (cell == null || cell === '') return null;
  const cleaned = String(cell).replace(',', '.').replace(/[^\d.-]/g, '');
  const n = parseFloat(cleaned);
  return Number.isFinite(n) ? n : null;
}

/** Zeitpunkt aus einer Zelle — mehrere Schreibweisen und reine Zeitstempel. */
function toDate(cell) {
  if (!cell) return null;
  const raw = String(cell).trim();

  // Reiner Zeitstempel in Sekunden oder Millisekunden
  if (/^\d+$/.test(raw)) {
    const n = parseInt(raw, 10);
    return n > 100000000000 ? n : n * 1000;
  }

  // 01.08.2026 oder 01.08.2026 07:30
  let m = raw.match(/^(\d{1,2})\.(\d{1,2})\.(\d{4})(?:[ T](\d{1,2}):(\d{2})(?::(\d{2}))?)?$/);
  if (m) return new Date(+m[3], +m[2] - 1, +m[1], +(m[4] || 0), +(m[5] || 0), +(m[6] || 0)).getTime();

  // 2026-08-01 oder 2026-08-01T07:30:00
  m = raw.match(/^(\d{4})-(\d{1,2})-(\d{1,2})(?:[ T](\d{1,2}):(\d{2})(?::(\d{2}))?)?/);
  if (m) return new Date(+m[1], +m[2] - 1, +m[3], +(m[4] || 0), +(m[5] || 0), +(m[6] || 0)).getTime();

  // 08/01/2026 — im Zweifel Monat zuerst, wie in den meisten Exporten
  m = raw.match(/^(\d{1,2})\/(\d{1,2})\/(\d{4})(?:[ T](\d{1,2}):(\d{2}))?$/);
  if (m) return new Date(+m[3], +m[1] - 1, +m[2], +(m[4] || 0), +(m[5] || 0)).getTime();

  const parsed = Date.parse(raw);
  return Number.isFinite(parsed) ? parsed : null;
}

/**
 * Liest den Inhalt einer geteilten Datei.
 *
 * Erwartet eine Kopfzeile mit Spaltennamen und danach eine Zeile je Messung.
 * Liefert `{ measurements, skipped, error }`.
 */
export function parseBodyFile(text, now = Date.now()) {
  const lines = String(text)
    .split(/\r?\n/)
    .map((l) => l.trim())
    .filter((l) => l.length > 0);

  if (lines.length < 2) {
    return { measurements: [], skipped: 0, error: 'Die Datei enthält keine Messungen.' };
  }

  const sep = detectSeparator(lines[0]);
  const header = splitLine(lines[0], sep).map((c) => c.toLowerCase());

  const iWeight = indexOfAny(header, WEIGHT);
  if (iWeight < 0) {
    return {
      measurements: [], skipped: 0,
      error: 'In der Kopfzeile ist keine Gewichtsspalte zu finden. '
        + `Gefunden wurde: ${header.join(', ').slice(0, 120)}`,
    };
  }

  const iDate = indexOfAny(header, DATE);
  const iFat = indexOfAny(header, FAT);
  const iLean = indexOfAny(header, LEAN);
  const iWater = indexOfAny(header, WATER);
  const iBone = indexOfAny(header, BONE);
  const iBasal = indexOfAny(header, BASAL);

  /** Nimmt den Wert nur, wenn er im plausiblen Bereich liegt. */
  const inRange = (v, lo, hi) => (v != null && v >= lo && v <= hi ? v : null);

  const out = [];
  let skipped = 0;

  lines.slice(1).forEach((line) => {
    const cells = splitLine(line, sep);
    const weight = toNumber(cells[iWeight]);
    // Werte außerhalb dieses Bereichs sind keine Körpergewichte
    if (weight == null || weight < 25 || weight > 300) {
      skipped++;
      return;
    }
    const at = (iDate >= 0 ? toDate(cells[iDate]) : null) ?? now;
    const basal = iBasal >= 0 ? inRange(toNumber(cells[iBasal]), 500, 5000) : null;

    out.push({
      at,
      weightKg: weight,
      bodyFatPercent: iFat >= 0 ? inRange(toNumber(cells[iFat]), 1, 70) : null,
      leanMassKg: iLean >= 0 ? inRange(toNumber(cells[iLean]), 10, 200) : null,
      waterMassKg: iWater >= 0 ? inRange(toNumber(cells[iWater]), 5, 150) : null,
      boneMassKg: iBone >= 0 ? inRange(toNumber(cells[iBone]), 0.5, 10) : null,
      basalKcal: basal == null ? null : Math.round(basal),
    });
  });

  if (out.length === 0) {
    return { measurements: [], skipped, error: 'Es ließ sich keine gültige Messung herauslesen.' };
  }

  // Je Tag nur die letzte Messung behalten, wie beim Weg über Health Connect
  const byDay = new Map();
  out.forEach((m) => {
    const day = Math.floor(m.at / DAY_MS);
    const old = byDay.get(day);
    if (!old || m.at >= old.at) byDay.set(day, m);
  });

  return {
    measurements: [...byDay.values()].sort((a, b) => a.at - b.at),
    skipped,
    error: null,
  };
}

/** Führt eingelesene Messungen mit den vorhandenen zusammen, ohne Dubletten. */
export function mergeBody(existing, incoming) {
  const byDay = new Map();
  (existing || []).forEach((m) => byDay.set(Math.floor(m.at / DAY_MS), m));
  (incoming || []).forEach((m) => {
    const day = Math.floor(m.at / DAY_MS);
    const old = byDay.get(day);
    // Die jüngere Messung des Tages gewinnt
    if (!old || m.at >= old.at) byDay.set(day, m);
  });
  return [...byDay.values()].sort((a, b) => a.at - b.at);
}

File: android/app/src/main/java/at/rudeboy/ferratafit/data/BodyImport.kt
package at.rudeboy.ferratafit.data

import java.text.SimpleDateFormat
import java.util.Locale
import java.util.TimeZone

/**
 * Einlesen von Waagendaten aus einer geteilten Datei.
 *
 * Hintergrund: Ein direkter Zugriff auf die FitDays-App ist nicht möglich — Android
 * kapselt Apps gegeneinander ab, und FitDays bietet keine Schnittstelle an. Der Weg
 * über Health Connect ist der bequemere, setzt aber voraus, dass die Kette
 * FitDays → Samsung Health → Health Connect steht.
 *
 * Klappt das nicht, bleibt der Export: FitDays und die meisten Waagen-Apps können ihre
 * Daten als Tabelle teilen. Dieser Leser ist bewusst großzügig, weil jede App ein
 * eigenes Format verwendet — er sucht in der Kopfzeile nach passenden Spalten und
 * kommt mit Punkt wie Komma als Dezimaltrenner zurecht.
 */
object BodyImport {

    /** Was beim Einlesen herauskam. */
    data class Result(
        val measurements: List<BodyMeasurement>,
        val skipped: Int,
        val error: String? = null
    )

    /** Spaltennamen, die je Wert in Frage kommen — in mehreren Sprachen. */
    private val WEIGHT = listOf("gewicht", "weight", "masse", "kg")
    private val DATE = listOf("datum", "date", "zeit", "time", "gemessen", "measured")
    private val FAT = listOf("körperfett", "koerperfett", "fett", "body fat", "bodyfat", "fat")
    private val LEAN = listOf("magermasse", "muskel", "muscle", "lean")
    private val WATER = listOf("wasser", "water")
    private val BONE = listOf("knochen", "bone")
    private val BASAL = listOf("grundumsatz", "bmr", "basal", "kcal")

    private val DATE_FORMATS = listOf(
        "yyyy-MM-dd HH:mm:ss", "yyyy-MM-dd'T'HH:mm:ss", "yyyy-MM-dd HH:mm", "yyyy-MM-dd",
        "dd.MM.yyyy HH:mm:ss", "dd.MM.yyyy HH:mm", "dd.MM.yyyy",
        "MM/dd/yyyy HH:mm:ss", "MM/dd/yyyy HH:mm", "MM/dd/yyyy",
        "dd/MM/yyyy HH:mm", "dd/MM/yyyy"
    )

    /**
     * Liest den Inhalt einer geteilten Datei.
     *
     * Erwartet eine Kopfzeile mit Spaltennamen und danach eine Zeile je Messung.
     * Trennzeichen darf Komma, Semikolon oder Tabulator sein.
     */
    fun parse(text: String, now: Long = System.currentTimeMillis()): Result {
        val lines = text.lineSequence()
            .map { it.trim() }
            .filter { it.isNotBlank() }
            .toList()

        if (lines.size < 2) {
            return Result(emptyList(), 0, "Die Datei enthält keine Messungen.")
        }

        val separator = detectSeparator(lines[0])
        val header = split(lines[0], separator).map { it.lowercase().trim('"', ' ') }

        val iWeight = indexOfAny(header, WEIGHT)
        if (iWeight < 0) {
            return Result(
                emptyList(), 0,
                "In der Kopfzeile ist keine Gewichtsspalte zu finden. " +
                    "Gefunden wurde: ${header.joinToString(", ").take(120)}"
            )
        }
        val iDate = indexOfAny(header, DATE)
        val iFat = indexOfAny(header, FAT)
        val iLean = indexOfAny(header, LEAN)
        val iWater = indexOfAny(header, WATER)
        val iBone = indexOfAny(header, BONE)
        val iBasal = indexOfAny(header, BASAL)

        val out = mutableListOf<BodyMeasurement>()
        var skipped = 0

        lines.drop(1).forEach { line ->
            val cells = split(line, separator)
            val weight = number(cells.getOrNull(iWeight))
            // Werte außerhalb dieses Bereichs sind keine Körpergewichte
            if (weight == null || weight < 25.0 || weight > 300.0) {
                skipped++
                return@forEach
            }
            val at = iDate.takeIf { it >= 0 }?.let { date(cells.getOrNull(it)) } ?: now
            out += BodyMeasurement(
                at = at,
                weightKg = weight,
                bodyFatPercent = iFat.takeIf { it >= 0 }?.let { number(cells.getOrNull(it)) }
                    ?.takeIf { it in 1.0..70.0 },
                leanMassKg = iLean.takeIf { it >= 0 }?.let { number(cells.getOrNull(it)) }
                    ?.takeIf { it in 10.0..200.0 },
                waterMassKg = iWater.takeIf { it >= 0 }?.let { number(cells.getOrNull(it)) }
                    ?.takeIf { it in 5.0..150.0 },
                boneMassKg = iBone.takeIf { it >= 0 }?.let { number(cells.getOrNull(it)) }
                    ?.takeIf { it in 0.5..10.0 },
                basalKcal = iBasal.takeIf { it >= 0 }?.let { number(cells.getOrNull(it)) }
                    ?.takeIf { it in 500.0..5000.0 }?.toInt()
            )
        }

        if (out.isEmpty()) {
            return Result(emptyList(), skipped, "Es ließ sich keine gültige Messung herauslesen.")
        }

        // Je Tag nur die letzte Messung behalten, wie beim Weg über Health Connect
        val deduped = out
            .groupBy { it.at / (24 * 60 * 60 * 1000L) }
            .map { (_, sameDay) -> sameDay.maxBy { it.at } }
            .sortedBy { it.at }

        return Result(deduped, skipped)
    }

    /** Führt eingelesene Messungen mit den vorhandenen zusammen, ohne Dubletten. */
    fun merge(existing: List<BodyMeasurement>, incoming: List<BodyMeasurement>): List<BodyMeasurement> {
        val byDay = existing.associateBy { it.at / (24 * 60 * 60 * 1000L) }.toMutableMap()
        incoming.forEach { m ->
            val day = m.at / (24 * 60 * 60 * 1000L)
            val old = byDay[day]
            // Die jüngere Messung des Tages gewinnt
            if (old == null || m.at >= old.at) byDay[day] = m
        }
        return byDay.values.sortedBy { it.at }
    }

    private fun detectSeparator(header: String): Char = listOf(';', '\t', ',')
        .maxBy { sep -> header.count { it == sep } }

    private fun split(line: String, sep: Char): List<String> =
        line.split(sep).map { it.trim().trim('"') }

    private fun indexOfAny(header: List<String>, needles: List<String>): Int =
        header.indexOfFirst { cell -> needles.any { cell.contains(it) } }

    /** Zahl aus einer Zelle — akzeptiert Punkt wie Komma und ignoriert Einheiten. */
    private fun number(cell: String?): Double? {
        if (cell.isNullOrBlank()) return null
        val cleaned = cell.replace(',', '.').filter { it.isDigit() || it == '.' || it == '-' }
        return cleaned.toDoubleOrNull()
    }

    private fun date(cell: String?): Long? {
        if (cell.isNullOrBlank()) return null
        // Reiner Zeitstempel in Sekunden oder Millisekunden
        cell.trim().toLongOrNull()?.let {
            return if (it > 100_000_000_000L) it else it * 1000L
        }
        DATE_FORMATS.forEach { pattern ->
            runCatching {
                val fmt = SimpleDateFormat(pattern, Locale.GERMAN)
                fmt.timeZone = TimeZone.getDefault()
                fmt.isLenient = false
                return fmt.parse(cell.trim())?.time
            }
        }
        return null
    }
}
